File: examples/guests/cpp/encoder/encoder.cpp
#include "generated/guest/init.hpp"
#include <string>
#include <algorithm>

namespace polyplug_plugin {

class EncoderImpl : public PipelineEncoderGuestContract {
public:
    explicit EncoderImpl(const HostApi* host) : host_(host) {}

    StringView encode(StringView input) override {
        std::string_view sv = polyplug::abi::strip_prefix(input, "TRANSFORMED:");
        std::string s(sv);
        std::replace(s.begin(), s.end(), '|', ',');
        return polyplug::alloc_string(host_, s);
    }

private:
    // Host handle for this runtime, captured at instance creation.
    const HostApi* host_;
};

// Factory called by the generated create_instance for every host-created
// instance. Ownership of the returned object transfers to the instance.
PipelineEncoderGuestContract* polyplug_create_encoder(const HostApi* host) {
    return new EncoderImpl(host);
}

}  // namespace polyplug_plugin

File: examples/guests/csharp/encoder/Encoder.cs
using System.Runtime.CompilerServices;
using Polyplug.Guest;
using Polyplug.Abi;

namespace Encoder;

public sealed class EncoderPlugin : IPipelineEncoderGuestContract
{
    // Host handle for this runtime, captured at instance creation.
    private readonly IntPtr _host;

    public EncoderPlugin(IntPtr host)
    {
        _host = host;
    }

    public StringView Encode(StringView input)
    {
        string s = StringViewHelper.StripPrefix(input, "TRANSFORMED:");
        return PolyplugHost.AllocString(_host, s.Replace('|', ','));
    }
}

public static class Registration
{
    [ModuleInitializer]
    public static void Register()
    {
        // The factory receives the HostApi pointer per created instance.
        EncoderInterfaces.SetEncoderFactory(host => new EncoderPlugin(host));
    }
}
